src/librustc_typeck/check/pat.rs

```diff
@@ -1391,15 +1400,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
     /// Type check the length of an array pattern.
     ///
-    /// Return the length of the variable length pattern,
+    /// Returns the length of the variable length pattern,
     /// if it exists and there are no errors.
+    ///
+    /// In case the pattern has a fixed length and the array
+    /// length is not yet determined, returns the array type
+    /// with the inferred length.
     fn check_array_pat_len(
         &self,
         span: Span,
+        inner_ty: Ty<'tcx>,
+        arr_ty: Ty<'tcx>,
         slice: Option<&'tcx Pat<'tcx>>,
         len: &ty::Const<'tcx>,
         min_len: u64,
-    ) -> Option<u64> {
+    ) -> Option<PatLenOrArray<'tcx>> {
```

Comment:
```suggestion
    ) -> (Ty<'tcx>, Ty<'tcx>) {
```
First one represents the `slice_ty` (so `tcx.err` where there isn't one) and the second represents the inferred type (i.e. the `expected` binding where this method is called).

Reply:
that's a lot cleaner :+1: